content/posts/09-nibbler-flatpak.md

```diff
@@ -0,0 +1,230 @@
+        mv * /nibbler
+        mkdir -p /app
+        mv /nibbler /app
+        chmod +x /app/nibbler/nibbler
+```
+
+### .desktop file
+
+Nibbler itself has no .desktop file, so let's write one too, one can notice that
+`--no-sandbox` is passed, that is required because Flatpak doesn't allow SUID or
+root-owned binaries, so we can't use their sandbox.
+
+```ini
+[Desktop Entry]
+Name=Nibbler (Flatpak)
+Comment=Leela Chess Zero (Lc0) Interface
+Exec=/app/nibbler/nibbler --no-sandbox %U
+Icon=Nibbler
+StartupNotify=true
+Terminal=false
+Type=Application
+Categories=Game
+```
+
+### Packaging Lc0
+
+Now the hardest part, packaging the Neural Network Engine itself, nn-engines are
+composed of two components:
+
+1. binary which runs and calculate chess positions, providing value output used
+   by our Chess GUI, and it does so with the second component.
+2. a `weights' file, which is a collection of numbers and bytes connections that
+   somehow make the binary provide good output that makes the engine very very
+   strong.
+
+Lets package them, first by the dependencies of lc0, these can be found online
+as they are used in other Flatpak-ed packages:
+
+```yaml
+  - name: eigen
+    buildsystem: cmake-ninja
+    builddir: true
+    sources:
+     - type: archive
+       url: https://gitlab.com/libeigen/eigen/-/archive/3.3.8/eigen-3.3.8.tar.gz
+       sha256: 146a480b8ed1fb6ac7cd33fec9eb5e8f8f62c3683b3f850094d9d5c35a92419a
+  - name: openblas
+    no-autogen: true
+    make-args:
+      - DYNAMIC_ARCH=1
+      - FC=gfortran
+      - NO_LAPACKE=1
+      - USE_OPENMP=0 # OpenMP off by default, this hack skips 'test_fork' which crashes on i386
+    make-install-args:
+      - PREFIX=/app
+    sources:
+      - type: archive
+        url: https://github.com/xianyi/OpenBLAS/archive/v0.3.12.tar.gz
+        sha256: 65a7d3a4010a4e3bd5c0baa41a234797cd3a1735449a4a5902129152601dc57b
+```
+
+And now the binary and the neural network, the latter needs to be in the same directory
+as the binary and named nn, one can specify a specific network later but lets use the
+defaults:
+
+```yaml
+  - name: lc0
+    buildsystem: meson
+    config-opts:
+      - --buildtype=release
+      - -Dopenblas=true
+      - -Dopenblas_libdirs=/app/lib
+      - -Dgtest=false
+      - -Db_lto=true
+      - -Dopencl=false
+    sources:
+      - type: git
+        url: https://github.com/LeelaChessZero/lc0
+        tag: v0.26.3
+  - name: nn
+    buildsystem: simple
+    build-commands:
+      - |
+        set -e
+        mkdir -p /app/bin
+        mv nn /app/bin
+    sources:
+      - type: file
+        dest-filename: nn
+        url: https://training.lczero.org/get_network?sha=4df05b0f0e80523018c073fd151ba26d955140ba303e84cebd96e027c6e06a3e
+        sha256: 821aabc4316d49a526ba737a218fb868dd7c72c5e7b1a91eb8e6c805e3503028
+```
+
+This is good enough, but let's go further.
+
+### Packaging a NNUE Chess Engine
+
+There is a different Neural Network type than the one used by lc0, called
```

Comment:
lc0 only has that Neural Network type, we wouldn't need to package a different chess engine if we could just flip a switch and use NNUE

Reply:
```suggestion
There is a different Neural Network type than the one used by lc0 by default, called
```